### PR TITLE
MM-59271 Add DataLoader classes and batch loading missing users and statuses

### PR DESCRIPTION
**Pull request by hmhealey**
#### Summary
This PR builds on Zubair's work in https://github.com/mattermost/mattermost/pull/26644 by genericizing that logic and making it so that it's easier for other code to batch and dedupe requests for data and then making it so that all calls to the `getMissingProfilesByIds` and `getMissingProfilesByUsernames` use similar behaviour.

The existing actions and the `getMissingProfilesByX` actions now both instantiate a `DataLoader` class which handles managing the timeout or recurring interval to make those requests as well as all the batching logic.

Those behave slightly differently than each other. The code in `status_profile_polling.ts` is on the same 3 second interval as it used before because it's intended to load background data, and the `getMissingProfilesByX` waits up to 100ms to send a request because it's generally used by components and UI elements where the quicker response time seems important. It might be a good idea to have these use the same method and to increase the speed of the `status_profile_polling` interval, but I want to confirm that this works before making more changes.

There's a lot going on here, and I'm trying some new things, so I'll definitely explain more in some comments below

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-59271

#### Release Note
```release-note
Batched requests made by certain components for loading users and their statuses from the server
```

### CODE REVIEW

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (hmhealey) [inline webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:183]: Waiting for a `DataLoader` to load is really tricky because:
1. When requests are batched, you can have multiple promises waiting on a single promise. It's the opposite of `Promise.all` because we have multiple places waiting for a single piece of code to run instead of one piece of code waiting for multiple promises to resolve. We have to hold onto each of those promise's `resolve` functions to be able to archive that. 
2. When IDs are deduped or split across multiple requests due to paging, we need to wait until all of the IDs for a given promise have been loaded. I'm a bit worried about the performance implications of storing the IDs for each request, and that's another reason that I made the default version of `queueForLoading` non-blocking.

I spent a lot of time looking at this to make sure that the `DataLoader`'s state is updated "atomically" by making sure that the JS thread can't be pre-empted between the timeout starting and a batch of IDs being removed from the `DataLoader`'s internal state, so please let me know if you see anything that you're at all suspicious about

**Comment by hmhealey** [comment]
@M-ZubairAhmed @lieut-data @calebroseland I'm expecting that this review might take a bit longer because I'm doing some tricky stuff because this is heavily async and I'm trying to make it generic and reusable. Please let me know if you're too busy to review it now or if you'd like me to walk through the code at all since I want to make the review less painful if possible

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (lieut-data) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:151]: Waiting feels like the most complicated part of the data loader, and seem to just entrench the existing data model practices instead of moving us away from that temporal coupling.

Building on your comment above re: most places don't wait for results, could we explore removing the need for callers to wait and solving the unit testing another way?

File: webapp/channels/src/packages/mattermost-redux/src/store/configureStore.ts
Comment (lieut-data) [reply webapp/channels/src/packages/mattermost-redux/src/store/configureStore.ts:54]: > I don't want to have to manually reset the global DataLoaders after each test.

I'm wondering how it would look to just do this reset vs. add this conceptual complexity? While cumbersome, it would be immediately clear when reviewing other tests what's going on with the singleton vs. having it buried here, especially when debugging.

Also, is this problem only when writing tests for the data loader itself, or writing tests for the various dispatched thunks?

File: webapp/channels/src/packages/mattermost-redux/src/actions/users.ts
Comment (lieut-data) [inline webapp/channels/src/packages/mattermost-redux/src/actions/users.ts:193]: Why do the callers need to know about the `missingIds`?

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (lieut-data) [inline webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:81]: It took me a while to understand that this "interval" data loader didn't mean, "repeatedly (re)load the data on a given interval", but more akin to, "load this data once, just be slower about it."

Can we consolidate `IntervalDataLoader` and `DelayedDataLoader`?

File: webapp/channels/src/packages/mattermost-redux/src/actions/status_profile_polling.ts
Comment (lieut-data) [inline webapp/channels/src/packages/mattermost-redux/src/actions/status_profile_polling.ts:27]: It's unfortunate that plugins won't be able to take advantage of these semantics without updating their Redux library or copy/pasting all the supporting code for this thunk. Would we consider moving to a model where this action is a plain old javascript object, and the "things to load" are just stored in Redux itself? Then we could iterate and tinker on the implementation in the future (even switching to Sagas) without breaking the interface.

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (lieut-data) [inline webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:8]: I'm not a huge fan of abstract classes in this context. Is there a scenario where we consolidate this into a single class? (Especially if we can merge the interval and delayed data loaders together)?

(A Saga function would be even better, but I agree with using the tools we have for now ;P)

**Review by lieut-data — APPROVED** [review]
@hmhealey, I'm not living and breathing this code day-to-day, but I did have enough context from our previous conversations that made this review straight forward. I haven't requested changes for any of the below, and instead defer to you and the team to figure out next steps.

Let me know if I can help further!

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:24]: I don't think that's likely. We don't have anything specific in place to prevent a valid ID from being empty, but I have a feeling that there's a lot of logic that'd break down if an ID is empty. That said, I would be fine making callers responsible for passing valid identifiers.

@M-ZubairAhmed Any thoughts on this? I think I copied this part of the code from your code

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:174]: Good catch. That's left over from before I realized this needed to split these requests into multiple batches. I'll update this

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:202]: Yeah, I can do that. I wish there was an easier way to expose stuff only for testing, but I can make it protected and extend the class in the test to make it public. I wish JS had the "default" visibility that Java has since it would allow for this behaviour

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:151]: I do agree that removing that would be ideal, but apart from figuring out how to test this (something like Testing Library's `waitFor` seems like it would work), I think the biggest hurdle for that would be to make the code more resilient to missing information since one place we tend to wait for things like this is when we need related information for an object (like the author of a post). I can't really see how we'd get around that without some sort of synchronization.

This doesn't seem like something we can solve right now, but I'll look at making it more explicit if the caller is awaiting on the result like with `queueToLoad` vs `queueAndWait`. That might lead to code that's a bit repetitive, but it'll at least help differentiate those two cases enough that people should see more usage of the non-blocking `queueToLoad`

File: webapp/channels/src/packages/mattermost-redux/src/actions/users.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/actions/users.ts:193]: There's one place in the "user is typing" message component (`MsgTyping`) looks at the result to determine if a user was loaded so that it knows whether or not we've already updated their status. It really only cares about the number of items loaded, so returning the IDs is as good as returning the objects for it

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:81]: The naming isn't perfect, but I haven't been able to come up with anything more self explanatory. Maybe `PeriodicDataLoader` could work, but I don't think that's much better.

As for consolidating, I was considering that, but @M-ZubairAhmed had concerns about the performance impact of using repeated timeouts vs a recurring interval when the server and app are under really heavy load. Perhaps once this is in, we can revisit the ceiling tests to see how this would perform in both cases to see if it makes much of a difference

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (lieut-data) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:151]: > I think the biggest hurdle for that would be to make the code more resilient to missing information since one place we tend to wait for things like this is when we need related information for an object (like the author of a post).

@hmhealey, can you point me an example? At least for components, I was wondering if it could just handle the partial state, with the React component refreshing after the partial update to potentially trigger the second load. For non-component contexts, like dispatches, I wonder if we can get the information we need in a single API request instead?

File: webapp/channels/src/packages/mattermost-redux/src/actions/status_profile_polling.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/actions/status_profile_polling.ts:27]: That would be the idea with switching to Sagas since, once those are started, you just interact with those by dispatching plain actions. If we want to start using these in plugins, we could consider exposing these actions on the window or something similar, but I'd be hesitant about doing that until we've proven this system in the web app.

We're currently dealing with some issues because a web app change which we thought wouldn't affect anyone else broke the AI plugin, so I'm not wanting to encourage that for a bit 😅

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:8]: As long as those two cases are separate, I think this is fine as-is. I'd rather have two subclasses with their own polling behaviours than to have one class with two different polling behaviours mixed together

File: webapp/channels/src/packages/mattermost-redux/src/store/configureStore.ts
Comment (M-ZubairAhmed) [inline webapp/channels/src/packages/mattermost-redux/src/store/configureStore.ts:51]: May be we should add this file in CODEOWNERS as well

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (M-ZubairAhmed) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:24]: While under heavy load we observed that there is network request send without either null or empty status id. To counter that it was added before

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (M-ZubairAhmed) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:81]: Two are equivalent in both terms of server and end result i.e both both them in our case will fire after certain amount of time. I had run the same data loader of previous version with both setTimeout and setInterval and the latter one performed better. We can definetly see the diffrence once we load test to see quantify the difference.

**Review by M-ZubairAhmed — APPROVED** [review]

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (lieut-data) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:81]: Makes sense. Is there a reason to keep the `setTimeout` one at all if we know `setInterval` and fewer allocations are better?

Comment (M-ZubairAhmed) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:81]: I guess its better to quantify difference in both and then we can expire either one of them. At that time when i was working in the PR i just saw setTimeout was a bit poor but how much of that i didnt see and moved on with setInterval. So its better to validate again.

File: webapp/channels/src/packages/mattermost-redux/src/actions/users.ts
Comment (calebroseland) [inline webapp/channels/src/packages/mattermost-redux/src/actions/users.ts:177]: Would it be possible to encapsulate the `if` and `create` responsibilities to some sort of an `ensure...` function call? Or possibly reduce the number of config options repeated? 1/5

**Review by calebroseland — APPROVED** [review]
I like the overall direction this is going in, and think it'll improve the dedupe and batching problems we have. I do have a few concerns as far as boilerplate and scalability go, but as a stepping stone towards more scalable patterns, I think this is a great step. 

I do wonder if there's a way to combine the classes and make the fetch-timing mechanism configurable, or perhaps more of a functional approach and pipe it depending on the desired behavior.

Part of me wonders if adding a `priority?: 'low' | 'normal' | 'high'` parameter somewhere in a single pipeline would be more efficient than creating entirely separate `IntervalDataLoader` and `DelayedDataLoader` loaders for single pipelines.


The dedupe, batching, and caching topics here have me thinking of [RTK Query](https://redux-toolkit.js.org/rtk-query/overview) or possibly Apollo client with https://www.apollographql.com/docs/react/api/link/apollo-link-rest/  or https://www.apollographql.com/docs/apollo-server/data/fetching-rest as potential solutions down the road.

File: webapp/channels/src/packages/mattermost-redux/src/store/configureStore.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/store/configureStore.ts:54]: Yeah, the extra argument's type is part of the `ThunkAction` type that `ActionFunc` and friends are based off of. I intentionally didn't type it properly for now because I wanted to make sure that this approach doesn't cause any issues yet. It's part of the worrying that this is going to get used which has lead me to write multiple paragraphs explaining how this is used and why people should make extra sure they need it before using it 😅

I'd like to leave it as-is for now, but it'll be easy to fix once we've let it soak for a bit

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:151]: The places that we await on this for something other than an optimization are in a couple places in the threads actions:
- https://github.com/mattermost/mattermost/blob/master/webapp/channels/src/packages/mattermost-redux/src/actions/threads.ts#L71
- https://github.com/mattermost/mattermost/blob/master/webapp/channels/src/packages/mattermost-redux/src/actions/threads.ts#L141

In general though, there's a lot of places where we use this sort of "wait for new data" throughout a lot of the initial load and routing logic. A lot of that comes down a lot to the lack of data ownership and the overfetching that that requires like I documented in [that spec I wrote last year](https://docs.google.com/document/d/1jLFhV2P7NGafCYfR3UcX6z4JWBZEj4IKrl4nIxRJK3c/edit), so it'll be hard to remove the waiting until we have more infrastructure like `DataLoader` in place. Once code can load data more easily and more automatically, it'll be easier for it to wait for Redux to be populated rather than explicitly waiting on an action like it has has to do now.

As for getting more data in a single API request, I've proposed that in the past as "richer APIs", and GraphQL was moving in that direction as well, but that goes against the focus on more RESTful APIs that we've had since API v4.

All that said, I'm going to leave this as-is since, while supporting waiting is more complicated, the current code in this PR should be able to handle that as long as it has time to soak. Not waiting would be ideal, but it requires rethinking a lot of how we do things, and we need to try this out before we can commit to it

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:24]: We should probably look into what the root cause of that is at some point, but given we don't expect a zero or empty value to be a valid case, I'm going to leave this as-is

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:202]: I'm actually going to leave this as-is. Ideally, we wouldn't expose it, but because there's no easy way to only expose it for tests, I don't think it's worth the mental overhead to hide an arguably private method that'll be useless for whoever uses this

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:81]: In theory, `setTimeout` should be better during periods where there's no data being requested of a given type because there's nothing running while `setInterval` requires a number of running handlers periodically checking for when work is available. I doubt the differences between the two will be very apparent until we use this in a lot of places and we have the server under extreme load. If we find that there are notable differences between the two, perhaps we could even make the `DataLoader` smart enough to swap between them based on how many requests it's getting, but that's obviously way out of scope for an initial version.

We should continue this discussion, but I'm going to move on to QA reviews with this without changing the code, and we can figure out which one to get rid of as a followup. Having two versions is likely redundant, but given they're both written and thoroughly unit tested, reducing that to only one is more of an ideological concern.

I'm definitely very appreciative of this discussion though since it does mean that people really care about this logic and aren't just rubber stamp approving it though :)

File: webapp/channels/src/packages/mattermost-redux/src/store/configureStore.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/store/configureStore.ts:51]: Yeah, I can do that. It's probably good for us to know if anyone's adding any middleware or anything else we should know about

**Comment by hmhealey** [comment]
> I like the overall direction this is going in, and think it'll improve the dedupe and batching problems we have. I do have a few concerns as far as boilerplate and scalability go, but as a stepping stone towards more scalable patterns, I think this is a great step.
>
> I do wonder if there's a way to combine the classes and make the fetch-timing mechanism configurable, or perhaps more of a functional approach and pipe it depending on the desired behavior.

Batching these requests is the best reason/excuse I've found to start making a bit of progress in the direction of the data manager stuff I've wanted to work on for ages. It's definitely a stepping stone towards hooks like `useUser` and such which can handle loading data more elegantly and automatically.

The `DataLoader` classes should hopefully reduce some of the boilerplate as, outside of the somewhat arbitrary decision about how often to fetch, it does mean that new places where we want to use these just require a `fetchBatch` and a `batchSize` and they'll be good to go. We can even look to make something like `bindClientFunc` which can instantiate the loader automatically and then dispatch the results to Redux, but that's a future improvement. I very intentionally made the `DataLoader` class completely unaware of Redux which could add a bit of boilerplate, but I think it's an acceptable amount.

> Part of me wonders if adding a `priority?: 'low' | 'normal' | 'high'` parameter somewhere in a single pipeline would be more efficient than creating entirely separate `IntervalDataLoader` and `DelayedDataLoader` loaders for single pipelines.

Yeah, that's something I'd love to do, but that's a future feature. This is currently an MVP version of `DataLoader`.

Going more in the data manger direction, I also have other ideas about making the `DataLoader` aware of what data is loaded and when so that it could even handle automatic refetching of things which are stale. We sync so much via the websocket right now, and I think it would improve performance if we were to make less of those real time and rely on refetching data periodically to make them seemingly real time but without the "10000 people need to immediately know I updated my profile picture" behaviour we can run into now.

> The dedupe, batching, and caching topics here have me thinking of [RTK Query](https://redux-toolkit.js.org/rtk-query/overview) or possibly Apollo client with https://www.apollographql.com/docs/react/api/link/apollo-link-rest/ as potential solutions down the road.

They've got the same goals, but I'm trying to fit those ideas into our code base in a way that doesn't involve however long it would take to rewrite everything. If we were starting from scratch, those would be good options, but with the sheer size of our code base plus whatever plugins are using, switching to one of those would take far more effort than we could spend on it. Doing it ourselves lets us do it piecemeal and adapt it to our API, and if we do it right, it doesn't preclude us from switching the internals to one of those later on

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:38]: I'm going to leave that one as-is since I like the parallels between the various "identifiers" names. I could see `nextIdentifiersBatch`, but then it's even wordier than the current one 😅

File: webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts:145]: Good idea. We're continually naming these methods better since `queue` is much better than `addIdsToLoad` or whatever it was when I first submitted this

File: webapp/channels/src/packages/mattermost-redux/src/actions/users.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/actions/users.ts:177]: I'll leave that as a future improvement. I like this having small a bit of friction to prevent it being used everywhere while we're still validating that it works

**Comment by hmhealey** [comment]
@yasserfaraazkhan This PR changes the code behind how a bunch of areas in the app load user profiles such as the "X has joined the channel" system messages, the user avatars below a thread, the "X is typing..." message below the post textbox, and the Compliance Monitoring pagein the system console. The code I changed is shared though, so you don't need to check all of them individually. I'm also hoping at least some of those are covered by E2E tests as well

**Review by yasserfaraazkhan — APPROVED** [review]

**Pull request merged**
